File: safety_node/scripts/safety_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
import time
import numpy as np
# TODO: include needed ROS msg type headers and libraries
from sensor_msgs.msg import LaserScan
from nav_msgs.msg import Odometry
from ackermann_msgs.msg import AckermannDriveStamped, AckermannDrive
from rclpy.qos import qos_profile_sensor_data

class SafetyNode(Node):
    """
    The class that handles emergency braking.
    """
    def __init__(self):
        super().__init__('safety_node')
        """
        One publisher should publish to the /drive topic with a AckermannDriveStamped drive message.

        You should also subscribe to the /scan topic to get the LaserScan messages and
        the /ego_racecar/odom topic to get the current speed of the vehicle.

        The subscribers should use the provided odom_callback and scan_callback as callback methods

        NOTE that the x component of the linear velocity in odom is the speed
        """
        self.speed = 0.
        
        # TODO: create ROS subscribers and publishers.
        self.cb_times = []
        self.max_samples = 100
        self.declare_parameter('ttc_threshold', 1.0)
        self.drive_publisher = self.create_publisher(
            AckermannDriveStamped, 
            '/drive', 
            10)
        
        self.scan_subscription = self.create_subscription(
            LaserScan,
            '/scan',
            self.scan_callback,
            qos_profile_sensor_data)
        
        self.odom_subscription = self.create_subscription(
            Odometry,
            '/ego_racecar/odom',
            self.odom_callback,
            10)
           
        

    def odom_callback(self, odom_msg):
        # TODO: update current speed
        self.speed = odom_msg.twist.twist.linear.x

    def scan_callback(self, scan_msg):
        # TODO: calculate TTC
        start = time.perf_counter()
        # 1. Extract ranges and angles from the scan_msg
        ranges = np.array(scan_msg.ranges, dtype=np.float32)
        # calculate the angles for each beam: angle_i = angle_min + i * angle_increment
        # angles = scan_msg.angle_min + np.arange(len(ranges)) * scan_msg.angle_increment
        angles = scan_msg.angle_min + np.arange(len(ranges), dtype=np.float32) * scan_msg.angle_increment

        # 2. Calculate r_dot
        r_dot = - self.speed * np.cos(angles)

        # 3. Calculate TTC 
        ttc_array = np.full_like(ranges, np.inf, dtype=np.float32)
        valid = np.isfinite(ranges) & (ranges > 0.0)
        approaching = valid & (r_dot < 0.0)

        ttc_array[approaching] = ranges[approaching] / (-r_dot[approaching])

        # 4. Threshold TTC and decide whether to brake
        ttc_threshold = self.get_parameter('ttc_threshold').get_parameter_value().double_value
        min_ttc = np.min(ttc_array)
        # TODO: publish command to brake
        if self.speed > 0.1 and min_ttc < ttc_threshold:
            drive_msg = AckermannDriveStamped()
            drive_msg.header.stamp = self.get_clock().now().to_msg()
            drive_msg.drive.speed = 0.0
            drive_msg.drive.steering_angle = 0.0
            # drive_msg.drive.acceleration = -5.0  
            self.drive_publisher.publish(drive_msg)
            self.get_logger().warn(f"Emergency braking triggered. TTC: {min_ttc:.2f}s")
        
        # Calculate the execution time, then compare with C++ version
        end = time.perf_counter()
        dt_ms = (end - start) * 1000.0
        self.cb_times.append(dt_ms)

        if len(self.cb_times) >= self.max_samples:
            avg = sum(self.cb_times) / len(self.cb_times)
            worst = max(self.cb_times)
            self.get_logger().info(
                f"[scan_callback] avg={avg:.3f} ms, worst={worst:.3f} ms"
            )
            self.cb_times.clear()
def main(args=None):
    rclpy.init(args=args)
    safety_node = SafetyNode()
    rclpy.spin(safety_node)
    safety_node.get_logger().info("Shutting down safety node.")
    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    safety_node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: safety_node/src/safety_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"
#include <chrono>
#include <vector>
#include <numeric>
#include <std_msgs/msg/bool.hpp>
#include <algorithm>
#include <limits>
#include <cmath>

class Safety : public rclcpp::Node {

public:
    Safety() : Node("safety_node")
    {
        /*
        You should also subscribe to the /scan topic to get the
        sensor_msgs/LaserScan messages and the /ego_racecar/odom topic to get
        the nav_msgs/Odometry messages

        The subscribers should use the provided odom_callback and 
        scan_callback as callback methods

        NOTE that the x component of the linear velocity in odom is the speed
        */

        /// TODO: create ROS subscribers and publishers
        // parameter
        this->declare_parameter<double>("ttc_threshold", 0.5);
        this->get_parameter("ttc_threshold", ttc_threshold);

        // publisher: /drive
        drive_pub = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>(
            "/drive", 10);
        
        // subscriber: /ego_racecar/odom
        odom_sub = this->create_subscription<nav_msgs::msg::Odometry>(
            "/pf/pose/odom", 10,
            std::bind(&Safety::drive_callback, this, std::placeholders::_1));

        brake_pub = this->create_publisher<std_msgs::msg::Bool>("/aeb_brake", 10);

        // subscriber: /scan (sensor QoS)
        scan_sub = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", rclcpp::SensorDataQoS(),
            std::bind(&Safety::scan_callback, this, std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "C++ safety_node started");
        
    }

private:
    double speed = 0.0;
    double ttc_threshold = 1.0;
    /// TODO: create ROS subscribers and publishers
    std::vector<double> cb_times_ms_;
    size_t max_samples_ = 100;
    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr drive_pub;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_sub;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr brake_pub;

    bool brake_active = false;
    int clear_count = 0;

    double min_stop_time = 1.0;      // at least stop for 1 s
    int clear_confirm_count = 10;    // 10 scans to clear the obstacle
    rclcpp::Time brake_start_time;
    void drive_callback(const nav_msgs::msg::Odometry::ConstSharedPtr msg)
    {
        /// TODO: update current speed
        speed = msg->twist.twist.linear.x;
    }

    void scan_callback(const sensor_msgs::msg::LaserScan::ConstSharedPtr scan_msg) 
    {
        /// TODO: calculate TTC

        /// TODO: publish drive/brake message
                // keep parameter up to date (optional, but handy when you set it at runtime)
        auto t0 = std::chrono::steady_clock::now(); // start time
        this->get_parameter("ttc_threshold", ttc_threshold);
        // 1. Extract ranges and angles from the scan_msg
        const auto & ranges = scan_msg->ranges;
        const size_t n = ranges.size();
        if (n == 0) return;

        const double inf = std::numeric_limits<double>::infinity();

        double min_ttc = inf;

        // iterate over beams
        for (size_t i = 0; i < n; ++i) {
            const float r = ranges[i];

            // valid range check
            if (!std::isfinite(r) || r <= 0.0f) continue;

            const double angle = scan_msg->angle_min + static_cast<double>(i) * scan_msg->angle_increment;

            // 2. Calculate r_dot
            const double r_dot = -speed * std::cos(angle);
            
            // 3. Calculate TTC
            // if r_dot < 0: iTTC = r / (-r_dot) else +inf
            double ttc = inf;
            if (r_dot < 0.0) {
                ttc = static_cast<double>(r) / (-r_dot);
            }

            if (ttc < min_ttc) min_ttc = ttc;
        }

        // 4. Threshold TTC and decide whether to brake
        bool danger = (speed > 0.1 && min_ttc < ttc_threshold);

        if (danger) {
            if (!brake_active) {
                brake_start_time = this->now();
                RCLCPP_WARN(this->get_logger(),
                    "[AEB START] TTC=%.2f < %.2f, braking!",
                    min_ttc, ttc_threshold);
            }

            brake_active = true;
            clear_count = 0;
        } else {
            if (brake_active) {
                double stopped_time = (this->now() - brake_start_time).seconds();

                if (stopped_time >= min_stop_time) {
                    clear_count++;

                    if (clear_count >= clear_confirm_count) {
                        brake_active = false;
                        clear_count = 0;

                        RCLCPP_INFO(this->get_logger(),
                            "[AEB CLEAR] obstacle clear, releasing brake");
                    }
                }
            }
        }

        // publish brake state every scan
        std_msgs::msg::Bool brake_msg;
        brake_msg.data = brake_active;
        brake_pub->publish(brake_msg);

        // optional: also publish direct stop command when braking
        if (brake_active) {
            ackermann_msgs::msg::AckermannDriveStamped drive_msg;
            drive_msg.header.stamp = this->now();
            drive_msg.drive.speed = 0.0;
            drive_msg.drive.steering_angle = 0.0;
            drive_pub->publish(drive_msg);

            RCLCPP_WARN(this->get_logger(),
                "[AEB STOP] TTC=%.2f | stopped=%s | clear_count=%d",
                min_ttc,
                brake_active ? "true" : "false",
                clear_count);
        }

        // end execution time measurement
        auto t1 = std::chrono::steady_clock::now();
        double dt_ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

        cb_times_ms_.push_back(dt_ms);
        if (cb_times_ms_.size() >= max_samples_) {
            double avg = std::accumulate(cb_times_ms_.begin(), cb_times_ms_.end(), 0.0) / cb_times_ms_.size();
            double worst = *std::max_element(cb_times_ms_.begin(), cb_times_ms_.end());
            RCLCPP_INFO(this->get_logger(), "[scan_callback] avg=%.3f ms, worst=%.3f ms (N=%zu)",
                        avg, worst, cb_times_ms_.size());
            cb_times_ms_.clear();
        }
    }



};
int main(int argc, char ** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Safety>());
    rclcpp::shutdown();
    return 0;
}
